support deno

vixalien opened the pull request:
<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**:

This PR allows `stylus` to work on `deno`, an alternative to Node.JS

**Why**:

Deno is gaining traction, and I personally use stylus, so I wanted to contrinue using stylus with Deno (after I switched to it from Node)

**How**:

The fix is two-steps:

1. Use `Object.setPrototypeOf` instead of the deprecated `__proto__` (which Deno doesn't support)
2. Fix a call in the `boolean` node that called the `Boolean` function, causing deno to run into an infinite loop as it was calling the constructor from the constructor in a recursive way.

**I THINK THIS IS OKAY, BUT PLEASE VERIFY**

**Checklist**:

- [ ] Documentation N/A
- [ ] Unit Tests N/A
- [x] Code complete
- [ ] Changelog

May I recommend moving to ES6 classes (would help) in the future, instead of modifying the prototype chain manually as this results in code reliability & performance issues.

vixalien committed:
fix deno running infinite loop

vixalien committed:
use `Object.setPrototypeOf` instead of `__proto__`

vixalien commented:
I forgot to add that this is a very minor change (in face of the user) as the API is unmodified and everything (including tests) works exactly as expected.

dependabot[bot] committed:
chore(deps): bump undici from 5.10.0 to 5.19.1 in /docs (#2758)

Bumps [undici](https://github.com/nodejs/undici) from 5.10.0 to 5.19.1.
- [Release notes](https://github.com/nodejs/undici/releases)
- [Commits](https://github.com/nodejs/undici/compare/v5.10.0...v5.19.1)

---
updated-dependencies:
- dependency-name: undici
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

iChenLei committed:
ci: update nodejs matrix test suites (#2765)

* ci: update nodejs matrix test suite

* Update ci.yml

peterhauke committed:
Correct typos and copy edit (#2764)

Fix capitalization of Node.js, CSS, Stylus (CSS is already plural so CSSs is not really correct)

Correct minor language errors

hooozen committed:
doc(js.md): supplement parameters description (#2780) (#2781)

taraspolovyi committed:
add support for CSS Values 4 viewport-relative units (#2775)

Co-authored-by: Taras Polovyi <taras.polovyi@grammarly.com>

vixalien commented:
anyone?

iChenLei commented:
Can you add github action ci for stylus with deno environment ?

iChenLei committed:
chore: update pull request template (#2784)

dependabot[bot] committed:
chore(deps): bump json5 from 2.2.1 to 2.2.3 in /docs (#2747)

Bumps [json5](https://github.com/json5/json5) from 2.2.1 to 2.2.3.
- [Release notes](https://github.com/json5/json5/releases)
- [Changelog](https://github.com/json5/json5/blob/main/CHANGELOG.md)
- [Commits](https://github.com/json5/json5/compare/v2.2.1...v2.2.3)

---
updated-dependencies:
- dependency-name: json5
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

iChenLei committed:
ci: add nodejs v20 test (#2791)

dependabot[bot] committed:
chore(deps): bump vite from 2.9.15 to 2.9.16 in /docs (#2799)

Bumps [vite](https://github.com/vitejs/vite/tree/HEAD/packages/vite) from 2.9.15 to 2.9.16.
- [Release notes](https://github.com/vitejs/vite/releases)
- [Changelog](https://github.com/vitejs/vite/blob/v2.9.16/packages/vite/CHANGELOG.md)
- [Commits](https://github.com/vitejs/vite/commits/v2.9.16/packages/vite)

---
updated-dependencies:
- dependency-name: vite
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

vixalien committed:
add deno test

vixalien committed:
add deno CI workflow

vixalien commented:
@iChenLei  ptal

iChenLei commented on lib/nodes/object.js:
```diff
@@ -29,7 +29,7 @@ var Object = module.exports = function Object(){
  */
 
-Object.prototype.__proto__ = Node.prototype;
+globalThis.Object.setPrototypeOf(Object.prototype, Node.prototype);
```
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis#browser_compatibility See `globalThis ` compatibility

vixalien committed:
move deno test to own dir to stop it from being run as a test

vixalien committed:
use this instead of globalThis (for compat)

vixalien commented:
@iChenLei fixed

vixalien committed:
fix `this.Object`pointing to custom `Object`

this solution works for:

- newer node versions (12+) and deno where `globalThis` is defined
- old node versions (where `global` is defined)

vixalien commented:
@iChenLei I believe stylus's dependencies do not support older versions of node anymore.

In the logs you can see `mocha` says

```
unsupported engine for mocha@9.2.2: wanted: {"node":">= 12.0.0"} (current: {"node":"10.24.1","npm":"6.14.12"})
```

also `@adobe/css-tools` causes the test to fail since it uses an optional chaining operator (`?.`) which works on node v14+.

The `mocha` warning can be ignored. but the `@adobe/css-tools` one cannot be ignored. we'll need to use an older version of this package if supporting older versions of node is a must.

vixalien committed:
downgrade `@adobe/css-tools` to ensure node compatibility

vixalien commented:
@iChenLei ptal

iChenLei commented:
I have a question: why not directly convert `function` into `class` (only needing [Node.js v4+](https://node.green/#ES2015-functions-class)), instead of using `setPrototypeOf`? Plus, you don't need to care about the issues with `@adobe/css-tools`.

iChenLei reviewed (commented):
I have a question: why not directly convert `function` into `class` (only needing [Node.js v4+](https://node.green/#ES2015-functions-class)), instead of using `setPrototypeOf`? Plus, you don't need to care about the issues with `@adobe/css-tools`.

vixalien commented:
that can be done yes, I was surprised why stylus does this too. since it'll be a huge change, I'll be working on a separate PR. stay tuned

vixalien commented:
@iChenLei made a PR for Classes in #2813

iChenLei commented:
So, can we close this PR ?

vixalien commented:
yes.
